src/private/xml_parser.cpp

```diff
@@ -417,12 +418,95 @@ namespace adm {
       return audioBlockFormat;
     }
 
-    SpeakerPosition parseSpeakerPosition(std::vector<NodePtr> nodes) {
-      SpeakerPosition speakerPosition;
+
+    SpeakerPosition parseSpeakerPosition(
+        std::vector<NodePtr> nodes) {
+      std::vector<std::pair<NodePtr, CartesianCoordinateValue>>
+          cartesianCoordinates;
+      std::vector<std::pair<NodePtr, SphericalCoordinateValue>>
+          sphericalCoordinates;
+      for(auto const& element : nodes) {
+        auto coordinate = element->first_attribute("coordinate");
+        if(coordinate) {
+          auto axis = coordinate->value();
+          if(axis == std::string("X")
+              || axis == std::string("Y")
+              || axis == std::string("Z")) {
+            cartesianCoordinates.emplace_back(element, axis);
+          }
+          if(axis == std::string{"azimuth"}
+              || axis == std::string{"elevation"}
+              || axis == std::string{"distance"}) {
+            sphericalCoordinates.emplace_back(element, axis);
+          }
+        } else {
+          throw std::runtime_error("SpeakerPosition is missing coordinate attribute");
+        }
+      }
+
+      if(cartesianCoordinates.empty() && sphericalCoordinates.empty()) {
+        throw std::runtime_error("SpeakerPosition has neither cartesian nor spherical coordinates");
+      }
+
+      if(!cartesianCoordinates.empty() && !sphericalCoordinates.empty()) {
+        throw std::runtime_error("SpeakerPosition has both cartesian and spherical coordinates");
+      }
+
+      if(!cartesianCoordinates.empty()) {
+        return parseCartesianSpeakerPosition(cartesianCoordinates);
+      } else {
+        return parseSphericalSpeakerPosition(sphericalCoordinates);
+      }
+    }
+
+    CartesianSpeakerPosition parseCartesianSpeakerPosition(
+        const std::vector<std::pair<adm::xml::NodePtr, adm::CartesianCoordinateValue>>&
+        cartesianCoordinates) {
+      adm::CartesianSpeakerPosition speakerPosition;
+      adm::ScreenEdgeLock screenEdgeLock;
+      for (auto& coord : cartesianCoordinates) {
+        auto element = coord.first;
+        auto axe = coord.second;
+        auto bound = adm::xml::parseOptionalAttribute<adm::BoundValue>(element, "bound");
+        if (axe == "X") {
+          if (bound == boost::none) {
+            adm::xml::setValue<adm::X>(element, speakerPosition);
+            adm::xml::setOptionalAttribute<adm::HorizontalEdge>(element, "screenEdgeLock",
+                                                                screenEdgeLock);
+          } else if (bound.get() == "min") {
+            adm::xml::setValue<adm::XMin>(element, speakerPosition);
+          } else if (bound.get() == "max") {
+            adm::xml::setValue<adm::XMax>(element, speakerPosition);
+          }
```

Comment:
worth raising an error if the bound is neither min or max?

Reply:
Turns out that already happened via the validator on the Bound class. Have changed things slightly by catching the `InvalidString` the validator throws and throwing an `XmlParsingError` instead as it's slightly more helpful & reports the line number. Also added a test.